tpe/plugin/src/FreeGroupFeatures.cc

```diff
@@ -77,24 +77,56 @@ void FreeGroupFeatures::SetFreeGroupWorldPose(
   const Identity &_groupID,
   const PoseType &_pose)
 {
+  // The input _pose is the target world pose for the canonical link
+  // in the model! So we need to compute the world pose to set the model to
+  // so that the canonical link is placed at the specified _pose
+  tpelib::Link *link = nullptr;
   auto modelIt = this->models.find(_groupID.id);
-  auto linkIt = this->links.find(_groupID.id);
   if (modelIt != this->models.end())
   {
     if (modelIt->second != nullptr)
-      modelIt->second->model->SetPose(math::eigen3::convert(_pose));
+    {
+      tpelib::Entity &linkEnt = modelIt->second->model->GetCanonicalLink();
+      link = dynamic_cast<tpelib::Link *>(&linkEnt);
+    }
   }
-  else if (linkIt != this->links.end())
+  else
   {
-    if (linkIt->second != nullptr)
-      linkIt->second->link->SetPose(math::eigen3::convert(_pose));
+    auto linkIt = this->links.find(_groupID.id);
+    if (linkIt != this->links.end())
+    {
+      // assume canonical link
+      link = linkIt->second->link;
+    }
   }
-  else
+
+  if (!link)
   {
     ignwarn << "No free group with id [" << _groupID.id << "] found."
       << std::endl;
     return;
   }
+
+  math::Pose3d targetWorldPose = math::eigen3::convert(_pose);
+  math::Pose3d tfChange = targetWorldPose - link->GetWorldPose();
+
+  // get top level model
+  tpelib::Entity *parent = link->GetParent();
+  tpelib::Entity *model = nullptr;
+  while (parent && dynamic_cast<tpelib::Model *>(parent))
+  {
+    model = parent;
+    parent = model->GetParent();
+  }
+  if (!model)
+  {
+    ignerr << "No model for free group with [" << _groupID.id << "] found."
+      << std::endl;
+    return;
+  }
+
+  // set the pose
+  model->SetPose(tfChange * model->GetPose());
```

Comment:
Comparing with the existing code, I see that there was a code path where `link->SetPose` is called, but that's no longer the case. Is that intentional?

Reply:
yes that's intentional. TPE does not support moving a model by its canonical link so the idea here is to find the difference between the current pose of link and target pose and apply that transform to the model containing the link.